File: Reqnroll/Chapter19/19.2-SupplementaryAttachmentPattern/Before/WIMP.Specs/Support/Logging/AppLogContext.cs
using System.Collections.Concurrent;

namespace WIMP.Specs.Support.Logging;

/// <summary>
/// Scenario context to collect log entries from the hosted WIMP application.
/// The log entries can be saved to an attachment or can be used for diagnoses.
/// </summary>
public class AppLogContext
{
    private readonly ConcurrentQueue<string> logMessages = new();

    public IReadOnlyCollection<string> LogMessages => logMessages;

    public void AddLogMessage(string logMessage)
    {
        logMessages.Enqueue(logMessage);
    }

    public void SaveToFile(string outputPath)
    {
        File.WriteAllLines(outputPath, logMessages.ToArray());
    }
}

File: CucumberJVM/Chapter19/19.3-TestLogPattern/After/src/test/java/com/wimp/app/specs/support/logging/AppLogContext.java
package com.wimp.app.specs.support.logging;

import io.cucumber.spring.ScenarioScope;
import org.slf4j.event.Level;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Collection;
import java.util.Collections;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * Scenario context to collect log entries from the hosted WIMP application. The
 * log entries can be saved to an attachment or can be used for diagnosis.
 */
@Component
@ScenarioScope
public class AppLogContext {

    private final ConcurrentLinkedQueue<String> logMessages = new ConcurrentLinkedQueue<>();

    /**
     * Returns an unmodifiable view of the captured log messages.
     */
    public Collection<String> getLogMessages() {
        return Collections.unmodifiableCollection(logMessages);
    }

    public void addLogMessage(Level logLevel, String logMessage) {
        logMessages.add(logMessage);
    }

    public String saveToFile(String outputPath) {
        String logContent = String.join(System.lineSeparator(), logMessages);
        try {
            Files.writeString(Paths.get(outputPath), logContent);
        } catch (IOException e) {
            throw new RuntimeException("Failed to save application logs to " + outputPath, e);
        }
        return logContent;
    }
}
